### PR TITLE
Implement LocatedSpan::get_line().

### PR DESCRIPTION
**Pull request by kaj**
Add a function to get the full input line containing the (start point of the) LocatedSpan.

As suggested in #53.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -277,6 +277,45 @@ impl<T: AsBytes, X> LocatedSpan<T, X> {
...
+        let self_ptr = self_bytes.as_ptr();
+        let offset = self.get_column() - 1;
+        let the_line = unsafe {
+            assert!(
+                offset <= isize::max_value() as usize,
+                "offset is too big"
+            );
+            let line_start_ptr = self_ptr.offset(-(offset as isize));
+            slice::from_raw_parts(line_start_ptr, offset + self_bytes.len())
+        };
```
Comment (ProgVal) [inline src/lib.rs:315]: I feel like this code could/should be shared with `get_columns_and_bytes_before`. Could you add a private (unsafe) function for this?

**Review by ProgVal — CHANGES_REQUESTED** [review]
Thanks!

Could you add some tests for this, checking edge cases? (first/last line, beginning/end of line, etc.)

**Comment by kaj** [comment]
Yes, I'll write some tests and then I'll see if the unsafe code can be unified.

**Comment by ProgVal** [comment]
Actually, I'm not sure, but isn't `get_line()` just returning `get_columns_and_bytes_before().1`?

**Comment by kaj** [comment]
> Actually, I'm not sure, but isn't `the_line` equal to `get_columns_and_bytes_before().1`?

Not quite, as `the_line` may continue after the thing that is `get_columns_and_bytes_before().1`.   If you look at the (newly added) `line_for_non_ascii_chars` test, i think `get_columns_and_bytes_before().1` would be `"Förra raden var "`.

File: src/tests.rs [hunk 2]
```diff
@@ -414,3 +414,96 @@ fn it_should_display_hex() {
...
+         \nand a third";
+    assert_eq!(
+        StrSpan::new(data).slice(data.len()..).get_line(),
+        Some("and a third".as_ref()),
+    );
+}
+
+#[test]
+fn line_for_non_ascii_chars() {
+    // I don't really know if this Oriya text makes sense.
```
Comment (ProgVal) [inline src/tests.rs:492]: I don't think that's Oriya :p

**Review by ProgVal — COMMENTED** [review]
Does `get_line` never return `None`?

File: src/tests.rs (hunk 2 above)
Comment (kaj) [reply src/tests.rs:492]: Nope.  I decided to change it to a language I do know (my native Swedish), but forgot the comment.  :-)

**Comment by kaj** [comment]
> Does `get_line` never return `None`?

Actually, as I was writing tests, I found myself asking me the same question.  I don't think it does, the option is accidental.  I'll get rid of it.

**Comment by ProgVal** [comment]
> Not quite, as `the_line` may continue after the thing that is `get_columns_and_bytes_before().1`. If you look at the (newly added) `line_for_non_ascii_chars` test, i think `get_columns_and_bytes_before().1` would be `"Förra raden var "`.

Oh yeah, indeed

**Comment by kaj** [comment]
I've been afk most of the day, it's time for bed in my tz, and I'm not entirely sober, but I've attempted a refactoring of the two similar unsafe blocks to one.  Can look more at it in ten hours or so ...

File: src/lib.rs [hunk 3]
```diff
@@ -257,17 +257,24 @@ impl<T: AsBytes, X> LocatedSpan<T, X> {
         &self.fragment
     }
 
-    fn get_columns_and_bytes_before(&self) -> (usize, &[u8]) {
+    fn get_unoffsetted_slice(&self) -> &[u8] {
```
Comment (ProgVal) [inline src/lib.rs:260]: Hmm, the name doesn't make it clear what this does, but I can't think of a better one. :/
Also some comment to explain what it does would be good

**Review by ProgVal — COMMENTED** [review]
Don't worry, we're in no hurry.

I just realized that this code doesn't work if `LocatedSpan`s are sliced with a right bound.
For example, this fails:

```
#[test]
fn line_of_word_in_middle() {
    let data = "One line of text\
         \nFollowed by a second\
         \nand a third\n";
    assert_eq!(
        StrSpan::new(data)
            .slice(data.find('\n').unwrap()..data.find('\n').unwrap()+5)
            .get_line(),
        "Followed by a second".as_bytes(),
    );
}
```

because `self.get_unoffsetted_slice()` only returns the bytes before the `LocatedSpan` and the bytes in the `LocatedSpan` itself, but none of the bytes after, even if they are on the same line.

Unfortunately, I don't see a way out of this without including the size of the original `&[u8]` in every `LocatedSpan`; but it would increase the size of `LocatedSpan<_, ()>` from 20 to 28 bytes (probably 24 to 32 including the padding).
But this seems costly just to provide a convenience function that can already be implemented by users themselves with safe code.

What do you think?

File: src/lib.rs (hunk 3 above)
Comment (kaj) [reply src/lib.rs:260]: Yes, I considered `get_original_size`, as that is pretty much the intent.  But I didn't, since it has no way of reconstructing the original length.  So `unoffsetted` is what it actually does, it undos the offset.

As for the larger question about a possibly missing trailer of `get_line()`; Yes, I was a bit worried about that myself, but decided that it's not a big problem in my main use-case of reporting parse errors.  In cases I can think of that involves marking an interval (and not just a position) of a line, I think I would have two `LocatedSpan`s to combine, where each of them would (probably) be "the rest of input from a starting point".

But I agree that this should be explained somehow, both in i a comment at `get_unoffsetted_slice` and in the docstring of `get_line`.  I'll try to write something.

**Comment by kaj** [comment]
On the other hand, the "original" adress and length won't change while the parser creates all those subslices that are `LocatedSpan`s.  So maybe it would make sense for a LocatedSpan to be a slice and a reference to the original slice (since it already borrows the actual data from the original slice, I think lifetimes should not be a problem).  The line and offset could be calculated on demand, instead of the other way around.  That way, a LocatedSpan should fit in 24 bytes (plus 8*sizeof(extra)/8).  Which should be eight bytes (line: u32 + alignment) _less_ than the current size.

But that is a quite big change ...

**Comment by ProgVal** [comment]
Well that's what I was thinking of, minus removing the `line` (which I don't think we should, because it makes `get_line` run in linear time instead)

**Comment by ProgVal** [comment]
What about adding a function (or method of `LocatedSpan`) that takes the original string as argument, to returns the line?

That way we don't make the struct larger *and* we still provide that feature (and as a bonus, it won't rely on unsafe code

**Comment by kaj** [comment]
> What about adding a function (or method of `LocatedSpan`) that takes the original string as argument, to returns the line?

That's what I do in rsass before starting to use `LocatedSpan` at all (in kaj/rsass#62).  For me, the whole point of using `LocatedSpan` is that I can get the line number, column, and the line itself from inside a parser function, where I don't have the original string.

**Comment by ProgVal** [comment]
Hmm yeah, good point. You could put a ref to the original string in the extra, but it would bloat `LocatedSpan` with duplicate data in your case.

**Comment by kaj** [comment]
For now, I think the docstring on `get_line()` is enough to handle the case where the LocatedSpan ends before the line.

Replacing the `self.offset` with a reference to the original data _may_ be a good idea, but I think the pros and cons of that should be considered independent of this PR.

**Comment by ProgVal** [comment]
Yeah, you're right. Could you just rename `get_line` to `get_line_beginning`, so we can introduce `get_line` later without it being a breaking change?

**Comment by kaj** [comment]
> Yeah, you're right. Could you just rename `get_line` to `get_line_beginning`, so we can introduce `get_line` later without it being a breaking change?

Ok, done.  Should I also squash the commits on this branch to one?

**Comment by ProgVal** [comment]
I can squash it on my end :)
